app/src/main/java/org/hitlabnz/sensor_fusion_demo/Compass3D.java

```diff
@@ -0,0 +1,310 @@
+package org.hitlabnz.sensor_fusion_demo;
+
+import static android.content.Context.SENSOR_SERVICE;
+import android.hardware.SensorManager;
+import com.badlogic.gdx.ApplicationAdapter;
+import com.badlogic.gdx.Gdx;
+import com.badlogic.gdx.Input;
+import com.badlogic.gdx.InputAdapter;
+import com.badlogic.gdx.Version;
+import com.badlogic.gdx.graphics.Color;
+import com.badlogic.gdx.graphics.GL20;
+import com.badlogic.gdx.graphics.PerspectiveCamera;
+import com.badlogic.gdx.graphics.Texture;
+import com.badlogic.gdx.graphics.VertexAttributes.Usage;
+import com.badlogic.gdx.graphics.g2d.BitmapFont;
+import com.badlogic.gdx.graphics.g2d.SpriteBatch;
+import com.badlogic.gdx.graphics.g3d.Environment;
+import com.badlogic.gdx.graphics.g3d.Material;
+import com.badlogic.gdx.graphics.g3d.Model;
+import com.badlogic.gdx.graphics.g3d.ModelBatch;
+import com.badlogic.gdx.graphics.g3d.ModelInstance;
+import com.badlogic.gdx.graphics.g3d.attributes.ColorAttribute;
+import com.badlogic.gdx.graphics.g3d.attributes.TextureAttribute;
+import com.badlogic.gdx.graphics.g3d.environment.PointLight;
+import com.badlogic.gdx.graphics.g3d.utils.ModelBuilder;
+import com.badlogic.gdx.math.MathUtils;
+import com.badlogic.gdx.math.Quaternion;
+import com.badlogic.gdx.math.Vector3;
+import com.badlogic.gdx.scenes.scene2d.Stage;
+import com.badlogic.gdx.scenes.scene2d.ui.Label;
+import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle;
+import com.badlogic.gdx.graphics.g3d.model.Node;
+import com.badlogic.gdx.scenes.scene2d.ui.Table;
+import com.badlogic.gdx.utils.ScreenUtils;
+import org.hitlabnz.sensorfusionlib.orientationProvider.*;
+import org.hitlabnz.sensorfusionlib.representation.*;
+
+public class Compass3D extends ApplicationAdapter {
+    private final String TAG = Compass3D.class.getName();
+    private final SensorSelectionActivity activity;
+    private final SensorManager sm;
+    private Environment lights;
+    private PerspectiveCamera cam;
+    private ModelBatch modelBatch;
+    private Model model;
+    private ModelInstance instance;
+    private Texture north, east, south, west, zenith, nadir;
+    private SpriteBatch batch;
+    private BitmapFont font;
+    private Stage stage;
+    private Label label;
+    private MatrixF4x4 rm;
```

Comment:
`rm` is not a good variable name. Please avoid abbreviations. As a general rule: Abbreviations require a significant mental load to be decoded and as developers read much more code than they write, having easy to read variable names is _always_ preferred.

Reply:
> `rm` is not a good variable name. Please avoid abbreviations. As a general rule: Abbreviations require a significant mental load to be decoded and as developers read much more code than they write, having easy to read variable names is _always_ preferred. 

Can be renamed to "providerRotationMatrix".